tensornetwork/sparse/lanczos.py

```diff
@@ -0,0 +1,309 @@
+# Copyright 2019 The TensorNetwork Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of TensorNetwork structure."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from tensornetwork.backends import backend_factory
+import inspect
+import warnings
+from typing import Any, List, Optional, Text, Tuple, Type, Callable
+import numpy as np
+Tensor = Any
+
+
+class LinearOperator:
+  """
+  A linear operator class. `LinearOperator` is used to provide a 
+  common interface to `eigsh` and other sparse solvers across 
+  all supported backends.
+  """
+
+  def __init__(self, matvec: Callable,
+               shape: Tuple[Tuple[int, ...], Tuple[int, ...]],
+               dtype: Type[np.number], backend: Text) -> None:
+    """
+    Initialize a `LinearOperator`
+    Args:
+      matvec: A function which implements a matrix-vector multiplication.
+        Allowed signatures are: `matvec(x)`, `matvec(x, backend)`
+          and `matvec(x, backend=default)` with `default` 'numpy', 'tensorflow'
+          'pytorch', 'jax' or 'shell'.
+         `result = `matvec` should return `result` of the same type as `x` (i.e.
+          a `Tensor` object)
+          Note that any default value for argument `backend` of `matvec` is 
+          overridden
+          by the value of the `backend` argument of `LinearOperator.__init__`
+      shape: A tuple of outgoing and incoming shapes of the `LinearOperator`.
+        For example, if `matvec(x)` takes as input a tensor `x` of shape 
+        `x.shape = (2,3,4,5)` and returns `result` with 
+         `result.shape = (4,5,2)`, then `shape` has to be 
+         `shape = ((4,5,2), (2,3,4,5))`.
+      dtype: The data-type of the `LinearOperator`.
+      backend: The backend to be used with this `LinearOperator`. 
+        `backend` can be 'numpy', 'tensorflow', 'pytorch', 'jax' or 'shell'
+    Returns: 
+      None
+    Raises:
+      ValueError if `matvec` signature is other than the above mentioned.
+    """
+    self.shape = shape
+    self.backend = backend_factory.get_backend(backend, dtype)
+    args = inspect.getargspec(matvec)
```

Comment:
This is super sketchy and I don't like it. Why are you doing this?

Reply:
to be able to give a clearer error message in case the user violates the signature of `matvec`. I'll remove it.

Reply:
actually, also to see if `matvec` takes a `backend` argument or not.
Typically `matvec` will contain a `TensorNetwork` somewhere in its body, so a `backend` argument
is generally needed. However, `eigsh_lanczos` can still be used for more general `matvec` functions.
If for example one uses a `numpy` backend for the `LinearOperator`, then one can just use plain numpy
code in `matvec` with no reference to `TensorNetwork`. In this case an argument `backend` is somewhat 
out of place.